pkg/scheduler/framework/plugins/noderesources/balanced_allocation.go

```diff
@@ -67,37 +68,68 @@ func (ba *BalancedAllocation) ScoreExtensions() framework.ScoreExtensions {
+	}
+
 	return &BalancedAllocation{
 		handle: h,
 		resourceAllocationScorer: resourceAllocationScorer{
 			Name:                BalancedAllocationName,
 			scorer:              balancedResourceScorer,
-			resourceToWeightMap: defaultRequestedRatioResources,
+			resourceToWeightMap: resToWeightMap,
 			enablePodOverhead:   fts.EnablePodOverhead,
 		},
 	}, nil
 }
 
-// todo: use resource weights in the scorer function
 func balancedResourceScorer(requested, allocable resourceToValueMap) int64 {
-	cpuFraction := fractionOfCapacity(requested[v1.ResourceCPU], allocable[v1.ResourceCPU])
-	memoryFraction := fractionOfCapacity(requested[v1.ResourceMemory], allocable[v1.ResourceMemory])
-	// fractions might be greater than 1 because pods with no requests get minimum
-	// values.
-	if cpuFraction > 1 {
-		cpuFraction = 1
+	resourceToFraction := []float64{}
+	var totalFraction float64
+	for name, value := range requested {
+		// This could only happen if both requsted[resource] and allocable[resource] are zero, this case could be avoid by tainting
+		// the node so that the node with the scare resources could be untouched and leave more room for the pods which do need them.
+		if allocable[name] == 0 {
+			continue
+		}
+		fraction := fractionOfCapacity(value, allocable[name])
+		if fraction > 1 {
+			fraction = 1
+		}
+		totalFraction += fraction
+		resourceToFraction = append(resourceToFraction, fraction)
 	}
-	if memoryFraction > 1 {
-		memoryFraction = 1
+
+	std := 0.0
+
+	// For most cases, resource are limited to cpu and memory, the std could be simplified to std := (fraction1-fraction2)/2
+	// If there are only resource left, just set std to zero is enough.
+	if len(resourceToFraction) == 2 {
+		std = math.Abs((resourceToFraction[0] - resourceToFraction[1]) / 2)
+	} else if len(resourceToFraction) > 2 {
+		mean := totalFraction / float64(len(resourceToFraction))
+		var sum float64
+		for _, fraction := range resourceToFraction {
+			sum = sum + (fraction-mean)*(fraction-mean)
+		}
+		std = math.Sqrt(sum / float64(len(resourceToFraction)))
 	}
 
-	// Upper and lower boundary of difference between cpuFraction and memoryFraction are -1 and 1
-	// respectively. Multiplying the absolute value of the difference by `MaxNodeScore` scales the value to
-	// 0-MaxNodeScore with 0 representing well balanced allocation and `MaxNodeScore` poorly balanced. Subtracting it from
-	// `MaxNodeScore` leads to the score which also scales from 0 to `MaxNodeScore` while `MaxNodeScore` representing well balanced.
-	diff := math.Abs(cpuFraction - memoryFraction)
-	return int64((1 - diff) * float64(framework.MaxNodeScore))
+	// STD (standard deviation) is always a positive value. 1-deviation lets the score to be higher for node which has least deviation and
+	// multiplying it with `MaxNodeScore` provides the scaling factor needed.
+	return int64((1 - std) * float64(framework.MaxNodeScore))
 }
 
 func fractionOfCapacity(requested, capacity int64) float64 {
```

Comment:
`capacity == 0` is always false now. Just remove the function

Reply:
Done